releases/compiler/parallel.go

```diff
@@ -52,25 +45,36 @@ func (t *Parallel) Add(work Work) {
 	if work.Source == "" {
 		panic("work.Source not set")
 	}
+	// if work.Result == nil || *work.Result == "" {
+	// 	t.work <- work
+	// }
+}
 
-	if work.Result == nil || *work.Result == "" {
-		t.work <- work
+func (t *Parallel) Run(ctx context.Context, work ...Work) error {
+	workCh := make(chan Work, len(work))
+	for _, w := range work {
+		if w.Result != nil && *w.Result != "" {
+			continue
+		}
+		t.mustValidateWork(w)
+		workCh <- w
 	}
-}
+	close(workCh)
 
-func (t *Parallel) Run(ctx context.Context) error {
```

Comment:
The API design of Run was very confusing, and didn't respect the error group's context.

Previously - It depended on on work being added before hand to a channel via Add, and then it would run until the default case was reached when the channel would be open but have no work left on it, and it had case listening for context cancelations. The design of Add being separate to the Run feels like this API should be able to add work at any time to a running parallel compiler, and it can technically, but you never know if the run has already exited.

Proposed Change:
Drop the add. If we need to know the work upfront, we can give the work we have directly to Run. 
We create a channel locally, validate and add all the work to it, and close it immediately. Then we select over the work and context. If the context is canceled we exit gracefully, if there's no more work we exit the routine.
This way there's less churn when a context is canceled, and a parallel compiler can be used for multiple calls to Run safely.